kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js

```diff
@@ -30,22 +35,136 @@ grCollectionsPanel.controller('GrCollectionsPanelCtrl', [
         ctrl.error = true;
     }).catch(() => ctrl.error = true);
 
+    ctrl.selectedImages$ = selectedImages$;
 }]);
 
-grCollectionsPanel.controller('GrNodeCtrl', ['collections', function(collections) {
+grCollectionsPanel.factory('collectionsTreeState', ['$window', function($window) {
+    const localStorageKey = 'collectionsTreeState';
+    const defaultShowState = true;
+
+    function getStateFromStorage() {
+        const jsonStr = $window.localStorage.getItem(localStorageKey) || '{}';
+        return JSON.parse(jsonStr);
+    }
+
+
+    function setState(pathId, show) {
+        const newState = angular.extend({}, getStateFromStorage(), {[pathId]: show});
+        $window.localStorage.setItem(localStorageKey, JSON.stringify(newState));
+    }
+
+    function getState(pathId) {
+        const state = getStateFromStorage();
```

Comment:
Does this mean you're going to hit `localStorage` for every single node in the tree? Could be a bit expensive. Is it worth caching the tree once and for all (and invalidate/update the cache on `setState`)?